test/e2e/grpc_test.go

```diff
@@ -64,30 +64,42 @@ func hasPort(u string) bool {
 	return err == nil
 }
 
-func dial(host, domain string) (*grpc.ClientConn, error) {
+func dial(ctx *TestContext, host, domain string) (*grpc.ClientConn, error) {
+	defaultPort := "80"
+	if test.ServingFlags.HTTPS {
+		defaultPort = "443"
+	}
 	if !hasPort(host) {
 		host = net.JoinHostPort(host, defaultPort)
 	}
 	if !hasPort(domain) {
 		domain = net.JoinHostPort(domain, defaultPort)
 	}
 
-	if host != domain {
+	secureOpt := grpc.WithInsecure()
+	if test.ServingFlags.HTTPS {
+		tlsConfig := test.TLSClientConfig(context.Background(), ctx.t.Logf, ctx.clients)
+		tlsConfig.ServerName = strings.Split(domain, ":")[0] // Set ServerName for pseudo hostname with TLS.
+		secureOpt = grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig))
+	}
+
+	if host != domain && !test.ServingFlags.HTTPS {
 		// The host to connect and the domain accepted differ.
 		// We need to do grpc.WithAuthority(...) here.
 		return grpc.Dial(
 			host,
 			grpc.WithAuthority(domain),
 			grpc.WithInsecure(),
-			// Retrying DNS errors to avoid .xip.io issues.
+			// Retrying DNS errors to avoid .sslip.io issues.
 			grpc.WithDefaultCallOptions(grpc.WaitForReady(true)),
 		)
 	}
+
 	// This is a more preferred usage of the go-grpc client.
 	return grpc.Dial(
 		host,
-		grpc.WithInsecure(),
-		// Retrying DNS errors to avoid .xip.io issues.
+		secureOpt,
+		// Retrying DNS errors to avoid .sslip.io issues.
 		grpc.WithDefaultCallOptions(grpc.WaitForReady(true)),
 	)
 }
```

Comment:
Can we build up a list of `DialOption`s and perform a single dial call?

That would simplify all the if conditions. For example it's not clear to me if why we don't set the Authority when HTTPS is enabled and `host != domain`?

Reply:
Thank you. Updated.